source/common/ssl/ssl_socket.cc

```diff
@@ -227,6 +230,30 @@ std::string SslSocket::sha256PeerCertificateDigest() {
   return Hex::encode(computed_hash);
 }
 
+std::string SslSocket::peerCertificate() const {
+  bssl::UniquePtr<X509> cert(SSL_get_peer_certificate(ssl_.get()));
+  if (!cert) {
+    return "";
+  }
+
+  bssl::UniquePtr<BIO> buf(BIO_new(BIO_s_mem()));
+  RELEASE_ASSERT(buf != nullptr);
+  RELEASE_ASSERT(PEM_write_bio_X509(buf.get(), cert.get()) == 1);
+  BUF_MEM* bio_buf = nullptr;
+  BIO_get_mem_ptr(buf.get(), &bio_buf);
+  RELEASE_ASSERT(bio_buf != nullptr);
+  RELEASE_ASSERT(bio_buf->data != nullptr);
+  RELEASE_ASSERT(bio_buf->length != 0);
+  std::string pem = std::string(bio_buf->data, bio_buf->length);
+  // URL encoding shortcut
+  pem = std::regex_replace(pem, std::regex("\n"), "%0A");
```

Comment:
This and the following lines look potentially expensive to me.  Would I be better off iterating the string once and just switching on every character before pushing it onto the end of a second string?  I was worried a bit that char iteration could cause issues with different charsets, but I assume it'll be safe here since Im already dealing with a pem formatted string?